design/authenticators/authn-k8s/inject_client_cert_solution_design.md

```diff
@@ -0,0 +1,195 @@
+# Solution Design - Inject Client Cert fix
+[//]: # "Change the title above from "Template" to your design's title"
+
+## Table of Contents
+[//]: # "You can use this tool to generate a TOC - https://ecotrust-canada.github.io/markdown-toc/"
+
+## Useful links
+[//]: # "Add links that may be useful for the reader"
+
+- [Existing PR with solution](https://github.com/cyberark/conjur/pull/1789)
+  - In this solution we wait for the timeout to get an answer and if we didn't
+    get one we create another WebSocket with an "ls" command to verify that the
+    cert is injected.
+- [Known issue in k8s exec](https://github.com/kubernetes/kubernetes/issues/89899)
+
+## Background
+[//]: # "Give relevant background for the designed feature. What is the motivation for this solution?"
+We have an issue with the "inject client cert" process where we don't always return a correct response to the client regarding the injection success. The injection is done via a WebSocket to Kubectl Exec where we use a "tar" command and the cert content in STDIN. There is [an issue](https://github.com/kubernetes/kubernetes/issues/89899) in k8s where messages that are sent via WebSockets to Kubectl Exec are left hanging without returning a response when using STDIN. 
+
+## Issue description
+[//]: # "Elaborate on the issue you are writing a solution for"
+The current behaviour is that we close the WebSocket connection after we send the injection request to kubectl. This is not correct as this way we do not wait for kubectl to send its response.
+
+This has 2 implications:
+1. If the injection request has succeeded we will return a 200 OK response to the 'inject_client_cert' request even if the cert has not finished to be written to the pod. This can lead to a race-condition where the client will try to read the cert before it is there.
+1. If the injection request has failed in the k8s API, we will not write the error to the Conjur logs, and we will falsely return a 200 OK response to the 'inject_client_cert' request
+
+## Solution
+[//]: # "Elaborate on the solution you are suggesting in this page. Address the functional requirements and the non functional requirements that this solution is addressing. If there are a few options considered for the solution, mention them and explain why the actual solution was chosen over them. Add an execution plan when relevant. It doesn't have to be a full breakdown of the feature, but just a recommendation to how the solution should be approached."
```

Comment:
I appreciate the multiple options you've presented but it seems to me that options 1 and 3 are inferior to option 2. So why not just verifying the URL length issue?
This should be easy by sending a 8k cert, no?

If I understand it correctly - if the URL length is not an issue then option 2 will be chosen, no?

Reply:
Sounds the best option to me either

Reply:
i like presenting all the options that rose in the process so that the reader knows that they were considered. also - sometimes people have other perspective that adds to the doc and changes the decisions.

I added a sentence describing the order in which we will spike the options.